### DISCUSSION TITLE
ORM models and lazy loading. How to optionally prevent a field from being loaded?

### DISCUSSION THREAD

**Discussion by jamesrusso** [pydantic/pydantic#4508] [Question]
I have SQLAlchemy and am using Pydantic in ORM mode. There are some situations where I don't want every single field touched as Pydantic traverses my models. 

```
class Patient(BaseModel):
   patient_id: str 
   firstname: str
   notes: List[Note]

class Note(BaseModel): 
   node_id: str
   note: str
```

There are times I want to just query for a Patient without lazily loading (or even eager loading via join) all that patient notes. I just want to leave that data out of the results. 

I could create a new schema called PatientWithoutNotes, but then it just gets too klunky with all these schemas floating around. 

Any suggestions on the best way to handle this when using ORM mode? My thought it to use the raiseload() lazy scheme (to raise an exception when attempting to load a relationship) and then use a customer GetterDict to catch that exception and just return an empty list (or maybe a None).

```
class UserGetter(GetterDict):

    def get(self, key: str, default: Any) -> Any:
        try:
            return getattr(self._obj, key, default)
        except sqlalchemy.exc.InvalidRequestError: 
            return default
```

**Comment by NoahCardoza** [comment]
I doubt this is useful to you anymore but hopefully this helps someone else.

After contemplating your suggestion I looked into `sqlalchemy`'s code and came up with this. I'm not sure if this will work for every case, but it seem to work for my needs at the moment.

```py
from typing import Any, Optional

from pydantic import BaseModel, Field
from pydantic.utils import GetterDict
from sqlalchemy.orm.base import instance_dict

from .category import CategoryOut


class ORMNoLazyLoaderGetter(GetterDict):
    def get(self, key: str, default: Any) -> Any:
        try:
            return instance_dict(self._obj)[key]
        except KeyError:
            return default


class ProductOut(BaseModel):
    id: int = Field(..., description="The product's ID.", example=420)
    # ...
    category: Optional[CategoryOut] = Field(
        None, description="The an object describing the product's category.")

    class Config:
        orm_mode = True
        getter_dict = ORMNoLazyLoaderGetter
```

**Reply by jordantshaw** [reply]
this seems like an elegant solution if the requirements are fairly basic, and there is only one relationship that you dont want loaded.  But what happens when you have multiple relationships that you may or may not want loaded based off user input.  Any thoughts?

**Comment by axelmukwena** [comment]
@jamesrusso @jordantshaw could this help out with your issue?

I used `noload` option to explicitly tell sqlalchemy to not trigger a relationship load
https://github.com/pydantic/pydantic/issues/8192#issuecomment-1855502691

**Reply by jordantshaw** [reply]
Using the `noload` option causes the field to be returned as None. The problem with this is there is no way to determine if the field is actually a null value being returned from the database or if it was just set to null by `noload`.  

In my use case I needed to know if the field was unset or actually a valid null value.  I ended up having to create a custom SQLAlchemy serializer that dumps the query results into dictionaries, and then from there, loading it into the pydantic model.
